### PR TITLE
FIX client ip parse error in lua

### PR DESCRIPTION
**Pull request by scw00**
Hi @shukitchan 
 I've found that ts_fetch parse client_ip incorrectly.
```
264	      rc   = sscanf(addr, "%15s:%d", ipstr, &port);
(gdb) p addr
$1 = 0x40011cf8 "127.0.0.1:8080"
(gdb) n
265	      if (rc == 2) {
(gdb) p addr
$2 = 0x40011cf8 "127.0.0.1:8080"
(gdb) p port
$3 = 1073813776
(gdb) p ipstr
$4 = "127.0.0.1:8080\000\000\320k\350\361\377\177\000\000\245\277S\000\000\000\000"
(gdb) 
```

So maybe we can fix it!

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci]

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/469/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/459/

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/456/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/2152/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/2048/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/582/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/713/

**Comment by shukitchan** [comment]
good find. i see the problem there. will review it.

File: plugins/experimental/ts_lua/ts_lua_fetch.c [hunk 1]
```diff
@@ -260,9 +260,12 @@ ts_lua_fetch_one_item(lua_State *L, const char *url, size_t url_len, ts_lua_fetc
...
     if (lua_isstring(L, -1)) {
-      addr = luaL_checklstring(L, -1, &addr_len);
-      rc   = sscanf(addr, "%15s:%d", ipstr, &port);
-      if (rc == 2) {
+      addr          = luaL_checklstring(L, -1, &addr_len);
+      const char *p = strstr(addr, ":");
+      if (p && p - addr < 31) {
+        port = atoi(p + 1);
+        strncpy(ipstr, addr, p - addr);
+        ipstr[p - addr]     = 0;
```
Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua_fetch.c]: rc is a variable used below. So if rc is not 2, then a default will be used.

Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua_fetch.c]: so default is used all the time after this change

File: plugins/experimental/ts_lua/ts_lua_fetch.c [hunk 2]
```diff
@@ -260,9 +260,12 @@ ts_lua_fetch_one_item(lua_State *L, const char *url, size_t url_len, ts_lua_fetc
     lua_gettable(L, -2);
 
     if (lua_isstring(L, -1)) {
-      addr = luaL_checklstring(L, -1, &addr_len);
-      rc   = sscanf(addr, "%15s:%d", ipstr, &port);
-      if (rc == 2) {
+      addr          = luaL_checklstring(L, -1, &addr_len);
+      const char *p = strstr(addr, ":");
+      if (p && p - addr < 31) {
```
Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua_fetch.c]: what's the check for "p - addr < 31" for ? what does it mean?

File: plugins/experimental/ts_lua/ts_lua_fetch.c [hunk 3]
```diff
@@ -260,9 +260,12 @@ ts_lua_fetch_one_item(lua_State *L, const char *url, size_t url_len, ts_lua_fetc
...
+        port = atoi(p + 1);
```
Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua_fetch.c]: perhaps it is better to check addr_len as well in determining the port. i.e. use the whole string from p+1 to addr+addr_len

File: plugins/experimental/ts_lua/ts_lua_fetch.c [hunk 4]
```diff
@@ -260,9 +260,12 @@ ts_lua_fetch_one_item(lua_State *L, const char *url, size_t url_len, ts_lua_fetc
     lua_gettable(L, -2);
 
     if (lua_isstring(L, -1)) {
-      addr = luaL_checklstring(L, -1, &addr_len);
-      rc   = sscanf(addr, "%15s:%d", ipstr, &port);
```
Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua_fetch.c]: rc is a variable used in code block below. So if rc is not 2, then a default will be used.
therefore default is used all the time after this change.

**Comment by scw00** [comment]
updated!

File: plugins/experimental/ts_lua/ts_lua_fetch.c [hunk 5]
```diff
@@ -253,25 +253,30 @@ ts_lua_fetch_one_item(lua_State *L, const char *url, size_t url_len, ts_lua_fetc
...
   if (tb) {
     lua_pushlstring(L, "cliaddr", sizeof("cliaddr") - 1);
     lua_gettable(L, -2);
 
     if (lua_isstring(L, -1)) {
       addr = luaL_checklstring(L, -1, &addr_len);
-      rc   = sscanf(addr, "%15s:%d", ipstr, &port);
-      if (rc == 2) {
+      p    = strstr(addr, ":");
+      if (p && addr_len < 32) {
```
Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua_fetch.c]: now i understand where the 32 came from. I think this should be (p - addr < 32)

File: plugins/experimental/ts_lua/ts_lua_fetch.c [hunk 6]
```diff
@@ -253,25 +253,30 @@ ts_lua_fetch_one_item(lua_State *L, const char *url, size_t url_len, ts_lua_fetc
...
+        port = atoi(p + 1);
+        strncpy(ipstr, addr, p - addr);
+        ipstr[p - addr]     = 0;
         clientaddr.sin_port = htons(port);
         inet_aton(ipstr, (struct in_addr *)&clientaddr.sin_addr.s_addr);
```
Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua_fetch.c]: what if inet_aton() is invalid and returns 0? Can we take care of this case as well and use the default (i.e. set p = NULL)?

**Comment by scw00** [comment]
updated

**Comment by shukitchan** [comment]
[approve ci]

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/529/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/2222/

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/526/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/655/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/783/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/2118/

**Comment by shukitchan** [comment]
[approve ci]

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/539/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/552/

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/537/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/2233/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/666/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/2129/

**Comment by shukitchan** [comment]
interesting that it crashes in FreeBSD11 build.

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/795/

**Comment by shukitchan** [comment]
@zwoop you have any insight on why that can happen?

**Comment by shukitchan** [comment]
[approve ci]

**Comment by scw00** [comment]
Hmm! Failed again!! Do you have the permission to check the log on this CI env!

**Comment by scw00** [comment]
Hmm! Failed again!! Do you have the permission to check the log on this CI env!

**Comment by shukitchan** [comment]
No I don't. @PSUdaemon or @zwoop should have. can you guys help?

**Comment by shukitchan** [comment]
[approve ci]

**Comment by shukitchan** [comment]
in case it failed again, leif and phil said we may simply fix it by rebasing the PR.

**Comment by zwoop** [comment]
This looks like it succeeded all build jobs :-).

**Comment by scw00** [comment]
👍

**Pull request merged**